keps/sig-api-machinery/20190625-api-compression-to-ga.md

```diff
@@ -0,0 +1,141 @@
+---
+title: Graduate API gzip compression support to GA
+authors:
+  - "@smarterclayton"
+owning-sig: sig-api-machinery
+participating-sigs:
+  - sig-cli
+reviewers:
+  - "@lavalamp"
+  - "@liggitt"
+approvers:
+  - "@liggitt"
+  - "@lavalamp"
+editor: TBD
+creation-date: 2019-03-22
+last-updated: 2019-03-22
+status: implementable
+see-also:
+  - "https://github.com/kubernetes/kubernetes/issues/44164"
+replaces:
+superseded-by:
+---
+
+# Graduate API gzip compression to GA
+
+## Table of Contents
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Release Signoff Checklist](#release-signoff-checklist)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+    - [Non-Goals](#non-goals)
+  - [Proposal](#proposal)
+    - [1.15](#1-15)
+    - [1.16](#1-16)
+    - [1.17](#1-17)
+    - [Implementation Details](#implementation-details)
+    - [Risks and Mitigations](#risks-and-mitigations)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Version Skew Strategy](#version-skew-strategy)
+  - [Implementation History](#implementation-history)
+
+## Summary
+
+Kubernetes sometimes returns extremely large responses to clients outside of its local network, resulting in long delays for components that integrate with the cluster in the list/watch controller pattern. Kubernetes should properly support transparent gzip response encoding, while ensuring that the performance of the cluster does not regress for small requests.
+
+## Motivation
+
+In large Kubernetes clusters the size of protobuf or JSON responses may exceed hundreds of megabytes, and clients that are not on fast local networks or colocated with the master may experience bandwidth and/or latency issues attempting to synchronize their state with the server (in the case of custom controllers). Many HTTP servers and clients support transparent compression by use of the `Accept-Encoding` header, and support for gzip can reduce total bandwidth requirements for integrating with Kubernetes clusters for JSON by up to 10x and for protobuf up to 8x.
+
+### Goals
+
+Allow standard HTTP transparent `Accept-Encoding: gzip` behavior to work for large Kubernetes API requests, without impacting existing Go language clients (which are already sending that header) or causing a performance regression on the Kubernetes apiservers due to the additional CPU necessary to compress small requests.
+
+### Non-Goals
+
+* Support other compression formats like Snappy due to limited client support
+* Compress non-API responses
+* Compress watch responses
+
+## Proposal
+
+### 1.16
+
+* Update the existing incomplete alpha API compression to:
+  * Only occur on API requests
+  * Only occur on very large responses (>128KB)
+* Promote to GA directly since this is a standard feature of HTTP servers
```

Comment:
is there any behavioral difference in promoting to beta in 1.16 but enabling by default? that would be more in keeping with the "allow to disable for X releases" approach. assuming no problems surface, we could promote to GA and lock enabled in 1.17

Reply:
Probably fine to do beta.  Until we see the 5k numbers I’m not even positive there is a measurable high end impact, but no reason to rush